wrappers/rust/icicle-cuda-runtime/src/memory.rs

```diff
@@ -117,6 +134,66 @@ impl<T> HostSlice<T> {
         self.0
             .iter_mut()
     }
+
+    pub fn is_pinnable(&self) -> bool {
+        let pinnable = get_device_attribute(cudaDeviceAttr::cudaDevAttrHostRegisterSupported, 0).unwrap();
+        let lockable =
+            get_device_attribute(cudaDeviceAttr::cudaDevAttrPageableMemoryAccessUsesHostPageTables, 0).unwrap();
+
+        pinnable == 1 && lockable == 0
+    }
+
+    pub fn pin(&self, flags: CudaHostRegisterFlags) -> CudaResult<()> {
+        unsafe {
+            let ptr = self.as_ptr() as *mut c_void;
+            let flags_to_set = flags.bits();
+            cudaHostRegister(ptr, self.len(), flags_to_set as c_uint).wrap()
+        }
+    }
+
+    pub fn unpin(&self) -> CudaResult<()> {
+        unsafe {
+            let mut flags = 0;
+            let ptr = self.as_ptr() as *mut c_void;
+            cudaHostGetFlags(&mut flags, ptr).wrap()?;
+            cudaHostUnregister(ptr).wrap()
+        }
+    }
+
+    pub fn allocate_pinned(count: usize, flags: CudaHostAllocFlags) -> CudaResult<&'static mut Self> {
+        let size = count
+            .checked_mul(size_of::<T>())
+            .unwrap_or(0);
+        if size == 0 {
+            return Err(CudaError::cudaErrorMemoryAllocation); //TODO: only CUDA backend should return CudaError
+        }
+
+        let mut pinned_host_ptr = MaybeUninit::<*mut c_void>::uninit();
+
+        unsafe {
+            cudaHostAlloc(pinned_host_ptr.as_mut_ptr(), size, flags.bits).wrap()?;
+            let pinned_host_slice = from_raw_parts_mut(pinned_host_ptr.assume_init() as *mut T, count);
+            Ok(Self::from_mut_slice(pinned_host_slice))
+        }
+    }
+
+    pub fn free_pinned(&self) -> CudaResult<()> {
+        unsafe {
+            let mut flags: u32 = 0;
+            let ptr = self.as_ptr() as *mut c_void;
+            cudaHostGetFlags(&mut flags, ptr).wrap()?;
+            cudaFreeHost(ptr).wrap()
+        }
+    }
+
+    pub fn get_memory_flags(&self) -> CudaResult<u32> {
+        unsafe {
+            let mut flags: u32 = 1234;
```

Comment:
1234 is intentional?

Reply:
Technically it can be any value above 16 since the bitflags are 0,1,2,4,8 and it needs to be a uint. I'm fine changing it to another number that won't be misinterpreted as a set flag